src/main/java/de/tum/cit/aet/artemis/core/web/CourseResource.java

```diff
@@ -1453,4 +1456,35 @@ public ResponseEntity<Long> getNumberOfAllowedComplaintsInCourse(@PathVariable L
         long unacceptedComplaints = complaintService.countUnacceptedComplaintsByParticipantAndCourseId(participant, courseId);
         return ResponseEntity.ok(Math.max(complaintService.getMaxComplaintsPerParticipant(course, participant) - unacceptedComplaints, 0));
     }
+
+    /**
+     * GET courses/{courseId}/existing-exercise-details: Get the exercise names (and shortNames for {@link ExerciseType#PROGRAMMING} exercises)
+     * of all exercises with the given type in the given course.
+     *
+     * @param courseId     of the course for which all exercise names should be fetched
+     * @param exerciseType for which the details should be fetched, as the name of an exercise only needs to be unique for each exercise type
+     * @return {@link CourseExistingExerciseDetailsDTO} with the exerciseNames (and already used shortNames if a {@link ExerciseType#PROGRAMMING} exercise is requested)
+     */
+    @GetMapping("courses/{courseId}/existing-exercise-details")
+    @EnforceAtLeastEditorInCourse
+    public ResponseEntity<CourseExistingExerciseDetailsDTO> getExistingExerciseDetails(@PathVariable Long courseId, @RequestParam String exerciseType) {
+        log.debug("REST request to get details of existing exercises in course : {}", courseId);
+        Course course = courseRepository.findByIdWithEagerExercisesElseThrow(courseId);
+
+        Set<String> alreadyTakenExerciseNames = new HashSet<>();
+        Set<String> alreadyTakenShortNames = new HashSet<>();
+
+        boolean includeShortNames = exerciseType.equals(ExerciseType.PROGRAMMING.toString());
+
+        course.getExercises().forEach((exercise -> {
+            if (exercise.getType().equals(exerciseType)) {
+                alreadyTakenExerciseNames.add(exercise.getTitle());
+                if (includeShortNames && exercise.getShortName() != null) {
+                    alreadyTakenShortNames.add(exercise.getShortName());
+                }
+            }
+        }));
+
+        return ResponseEntity.ok(new CourseExistingExerciseDetailsDTO(alreadyTakenExerciseNames.toArray(String[]::new), alreadyTakenShortNames.toArray(String[]::new)));
```

Comment:
Why doesn't the dto directly contain the sets?

Reply:
Incomplete refactoring from previous review, I adjusted it